Introduce mono repo

choidabom opened the pull request:
<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it?

This PR proposes the implementation of a mono repo structure for the CodePair project, which currently consists of a React frontend and a NestJS backend. Adopting a monorepo will not only streamline our development process but also ensure consistency across the codebase, making it easier for developers to work collaboratively.



#### Any background context you want to provide?

Adopting a mono repo structure is beneficial as it simplifies the development workflow, improves consistency in dependency versions, and facilitates better cross-team collaboration. By aligning both the frontend and backend under a single repository, we aim to increase the efficiency and maintainability of the codebase.
Our sibling project, Yorkie-JS-SDK, is currently utilizing pnpm to manage its monorepo, which allows us to maintain consistency across projects.

#### What are the relevant tickets?
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes https://github.com/yorkie-team/codepair/issues/351

### Checklist
- [x] Added relevant tests or not required
- [x] Didn't break anything

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Transitioned package management from npm to pnpm for both frontend and backend applications.
	- Introduced a new pnpm workspace configuration for managing frontend and backend packages.

- **Documentation**
	- Updated README files for both frontend and backend to reflect the new pnpm commands and installation instructions.

- **Chores**
	- Cleaned up workflow files to enhance clarity and consistency in CI processes.
	- Added caching for pnpm in CI workflows to improve build efficiency.
	- Updated `.prettierignore` file to exclude generated files from formatting.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

devleejb commented on backend/Dockerfile:
It seems that this Dockerfile isn’t utilizing `pnpm-lock.yaml`.

We can refer to [Working with Docker | pnpm](https://pnpm.io/docker#example-2-build-multiple-docker-images-in-a-monorepo) as a guide.

devleejb commented on MAINTAINING.md:
```diff
@@ -4,8 +4,9 @@
-- Update `version` in `package-lock.json` in `frontend/` and `backend/`. It will be automatically updated when you run `npm install`.
+- For packages that are independently versioned (e.g., `frontend/`, `backend/`), update the `version` field in their respective `package.json` files when making a release.
+- Otherwise, centralize version management at the root `package.json` if using a unified versioning strategy.
+- The `version` in `pnpm-lock.yaml` will be automatically updated when you run `pnpm install` at the root of the mono repo.
```
It seems that the `version` doesn't exist in `pnpm-lock.yaml`.

devleejb commented on MAINTAINING.md:
```diff
@@ -4,8 +4,9 @@
 
-- Update `version` in [package.json in `frontend/`](https://github.com/yorkie-team/codepair/blob/1f10dff1d9c253f921ba136c812383fbb292078f/frontend/package.json#L5) and [package.json in `backend/`](https://github.com/yorkie-team/codepair/blob/1f10dff1d9c253f921ba136c812383fbb292078f/backend/package.json#L3).
-- Update `version` in `package-lock.json` in `frontend/` and `backend/`. It will be automatically updated when you run `npm install`.
+- For packages that are independently versioned (e.g., `frontend/`, `backend/`), update the `version` field in their respective `package.json` files when making a release.
+- Otherwise, centralize version management at the root `package.json` if using a unified versioning strategy.
```
CodePair uses global version management.
Can we manage the version only in `package.json`(not `frontend/package.json`, `backend/package.json`)?

When I write the version only in `package.json`, it doesn't detect `import.meta.env.PACKAGE_VERSION`.

It is just a question :)

choidabom replied:
As we don’t have to release each package separately like other monorepos, it should be okay to manage the version in the root package.json.

devleejb commented on .github/workflows/ci_backend.yaml:
It seems that this workflow isn’t utilizing `pnpm-lock.yaml`.

We can refer to [Continous Integration | pnpm](https://pnpm.io/continuous-integration#github-actions) as a guide.

devleejb commented on .github/workflows/ci_frontend.yaml:
It seems that this workflow isn’t utilizing `pnpm-lock.yaml`.

We can refer to [Continous Integration | pnpm](https://pnpm.io/continuous-integration#github-actions) as a guide.

devleejb commented on .github/workflows/gh_pages.yaml:
It seems that this workflow isn’t utilizing `pnpm-lock.yaml`.

We can refer to [Continous Integration | pnpm](https://pnpm.io/continuous-integration#github-actions) as a guide.

devleejb commented on pnpm-lock.yaml:
It seems that this file contains unused dependencies.
With this file, global `node_modules` becomes `1.4G`.

choidabom replied:
@devleejb How can I check node_modules size?

devleejb replied:
You can check it with `du -sh ./node_modules`, `du -sh ./backend/node_modules`, `du -sh ./frontend/node_modules`.

choidabom replied:
@devleejb How can I check node_modules size?

devleejb reviewed (changes requested):
After this PR is merged, we can discuss unifying the lint rules for the `frontend` and `backend`.

And, I checked about the disk saving.

### Disk Savings

`npm`
- Frontend: 409M
- Backend: 693M

`pnpm`
- Global: 983M
- Frontend: 28K
- Backend: 76K

We saved approximately 10% of disk space.

devleejb commented on backend/Dockerfile:
```diff
@@ -1,69 +0,0 @@
-    ca-certificates \
-    ttf-freefont
-
-# Download Korean font for Puppeteer
-RUN mkdir /usr/share/fonts/nanumfont && \
-    wget http://cdn.naver.com/naver/NanumFont/fontfiles/NanumFont_TTF_ALL.zip && \
-    unzip NanumFont_TTF_ALL.zip -d /usr/share/fonts/nanumfont && \ 
-    fc-cache -f -v
```
This part should be included in the 3rd stage of `Dockerfile` .
With this code, the export feature can support Korean.

choidabom replied:
I see. I think it was deleted incorrectly.

devleejb commented on .github/workflows/ci_frontend.yaml:
```diff
@@ -1,3 +1,5 @@
+# @format
+
```
What does this line mean?

devleejb commented on MAINTAINING.md:
```diff
@@ -1,11 +1,12 @@
 
-- Update `version` in [package.json in `frontend/`](https://github.com/yorkie-team/codepair/blob/1f10dff1d9c253f921ba136c812383fbb292078f/frontend/package.json#L5) and [package.json in `backend/`](https://github.com/yorkie-team/codepair/blob/1f10dff1d9c253f921ba136c812383fbb292078f/backend/package.json#L3).
-- Update `version` in `package-lock.json` in `frontend/` and `backend/`. It will be automatically updated when you run `npm install`.
+- CodePair uses global version management. Update the `version` field only in the root `package.json` file when making a release.
```
With global version management, the value of `import.meta.env.PACKAGE_VERSION` is undefined.
This value is injected by `vite-plugin-package-version` in `frontend/vite.config.ts`.

choidabom replied:
I think I was confused. But, I'm not sure how much detail to explain.

devleejb commented on backend/package.json:
The indentation is changed to space from tab.
In `.prettierrc`, we set `useTabs` to `true`.

choidabom replied:
I think it's because `type: "module"` is missing the backend package.json. And I'll modify the script so that `json` is properly formatted as well.

devleejb commented on backend/package.json:
```diff
@@ -1,104 +1,95 @@
 {
-	"name": "codepair-backend",
-	"version": "0.1.14",
```
It seems that we should add `version` field.

choidabom replied:
okay !

devleejb commented on frontend/package.json:
```diff
@@ -1,29 +1,27 @@
+	"name": "@codepair/frontend",
 	"description": "CodePair Frontend",
-	"private": true,
-	"version": "0.1.14",
```
It seems that we should add `version` field.

devleejb reviewed (changes requested):
It looks almost perfect! Thank you for your work.
I have a few comments:

1. It seems that backend docker image isn't running properly.
    <img width="1874" alt="image" src="https://github.com/user-attachments/assets/cbd105ac-5268-465f-ba2d-8f3bbaf080c3">
2. We should move `backend/.dockerignore` file to top directory.
3. Could you clarify what `# @format` means in the GitHub workflow files?

choidabom commented:
@devleejb Thank you for reviewing.

> It looks almost perfect! Thank you for your work. I have a few comments:
> 
> 1. It seems that backend docker image isn't running properly.
>    <img alt="image" width="1874" src="https://private-user-images.githubusercontent.com/52884648/375024961-cbd105ac-5268-465f-ba2d-8f3bbaf080c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NDc0NzQsIm5iZiI6MTcyODU0NzE3NCwicGF0aCI6Ii81Mjg4NDY0OC8zNzUwMjQ5NjEtY2JkMTA1YWMtNTI2OC00NjVmLWJhMmQtOGYzYmJhZjA4MGMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDEwVDA3NTkzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxMTVjMGMzMTA4NTdkYWEwNTE0YWI5NzQyZWU0MGVjOWE0ODdmZWMxYTc5ZWY1YTM5MWY3Y2U0YWQwNTQwMzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KuLKs4CW_hijo8CfnZjYLQCfRqnj27T0aZ2-h6sQqgk">
> 2. We should move `backend/.dockerignore` file to top directory.
> 3. Could you clarify what `# @format` means in the GitHub workflow files?

1. The main reason the backend Docker image failed to build was due to specifying `"type": "module"` in the package.json file. NestJS is designed to use CommonJS by default, so additional configuration is required when attempting to use ES modules.
On the other hand, there's no issue with specifying `"type": "module"` in the package.json file of the frontend folder. This is because Vite acts as a bundler, providing seamless support for ES module usage.
  - Two potential solutions can be considered for this situation:
  
    1. Remove the useTabs condition from the .prettierrc file to resolve configuration conflicts.
    2. Maintain `"type": "module"` in the backend, similar to the frontend, while applying the necessary additional configurations.


2. I changed it. 
3. It was added automatically by vscode. I deleted it.
---
Additionally, I added `RUN sed -i 's/"prepare": "husky install"/"prepare": ""/' ./package.json` in the Dockerfile. This command means replacing the "prepare": "husky install" script in the package.json file with an empty string, effectively disabling the Husky installation process.

The reason for adding this is because the image wasn't building properly, and removing the Husky setup helped resolve the issue. Additionally, since Git hooks aren't needed in the Docker environment, this change simplifies the build process.

devleejb reviewed (approved):
For now, it would be better to remove `type:module` and accept the conflict of `useTabs`.

Thank you for your contribution.
